src/proc_diskstats.c

```diff
@@ -69,21 +77,35 @@ struct disk *get_disk(unsigned long major, unsigned long minor) {
 		last->next = d;
 	}
 
+	// ------------------------------------------------------------------------
+	// find the type of the device
+
 	// find if it is a partition
-	// by reading /sys/dev/block/MAJOR:MINOR/partition
+	// by checking if /sys/dev/block/MAJOR:MINOR/partition is readable.
 	char buffer[FILENAME_MAX + 1];
-	snprintfz(buffer, FILENAME_MAX, path_find_block_device_partition, major, minor);
-
-	int fd = open(buffer, O_RDONLY, 0666);
-	if(likely(fd != -1)) {
-		// we opened it
-		int bytes = read(fd, buffer, FILENAME_MAX);
-		close(fd);
-
-		if(bytes > 0)
-			d->partition_id = strtoul(buffer, NULL, 10);
+	snprintfz(buffer, FILENAME_MAX, path_find_block_device, major, minor, "partition");
+	if(access(buffer, R_OK) == 0) {
+		d->type = DISK_TYPE_PARTITION;
+	} else {
+		// find if it is a container
+		// by checking if /sys/dev/block/MAJOR:MINOR/slaves has entries
+		snprintfz(buffer, FILENAME_MAX, path_find_block_device, major, minor, "slaves/");
+		DIR *dirp = opendir(buffer);	
+		if (dirp != NULL) {
+		struct dirent *dp;
+			while( (dp = readdir(dirp)) ) {
+				// . and .. are also files in empty folders.
+				if(strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0) {
+					continue;
+				}
+				d->type = DISK_TYPE_CONTAINER;
+				// Stop the loop after we found one file.
+				break;
```

Comment:
instead of breaking the loop, we could count the number of slaves and instead of `d->type` we could have `d->slaves`.

why?

We could use the number of slaves to decide if we should enable by default the virtual disk or not. Virtual disks with just one slave can be disabled by default, while virtual disks with more than one slave can be enabled by default.

Will this solve your virtual disks case while allowing RAID devices to be enabled by default?